cpp/src/groupby/hash/shared_memory_aggregator.cuh

```diff
@@ -0,0 +1,263 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cudf/detail/aggregation/aggregation.hpp>
+#include <cudf/detail/aggregation/device_aggregators.cuh>
+#include <cudf/detail/utilities/assert.cuh>
+#include <cudf/detail/utilities/device_atomics.cuh>
+#include <cudf/utilities/traits.cuh>
+
+#include <cuda/std/cstddef>
+#include <cuda/std/type_traits>
+
+namespace cudf::groupby::detail::hash {
+template <typename Source, cudf::aggregation::Kind k, typename Enable = void>
+struct update_target_element_shmem {
+  __device__ void operator()(cuda::std::byte* target,
+                             cudf::size_type target_index,
+                             bool* target_null,
+                             cudf::column_device_view source,
+                             cudf::size_type source_index) const
+  {
+    CUDF_UNREACHABLE("Invalid source type and aggregation combination.");
+  }
+};
+
+template <typename Source>
+struct update_target_element_shmem<
+  Source,
+  cudf::aggregation::MIN,
+  cuda::std::enable_if_t<cudf::is_fixed_width<Source>() && cudf::has_atomic_support<Source>()>> {
+  __device__ void operator()(cuda::std::byte* target,
+                             cudf::size_type target_index,
+                             bool* target_null,
+                             cudf::column_device_view source,
+                             cudf::size_type source_index) const noexcept
+  {
+    using DeviceTarget = cudf::detail::underlying_target_t<Source, aggregation::MIN>;
+    using DeviceSource = cudf::detail::underlying_source_t<Source, aggregation::MIN>;
+
+    DeviceTarget* target_casted = reinterpret_cast<DeviceTarget*>(target);
```

Comment:
How will this work for strings columns?
The target type will be a `string_view` and this cast will be incorrect.

Reply:
Good question. I wanted to give you a thorough response, so I took a deeper dive into the details to see how this works. The more I look into the current code, the more surprised I am that it's producing the correct results. 

Still working on it. Need more time to fully understand this.

Reply:
OK, finally got this. String columns don't have the atomic support so they use the sort-based groupby all the time.

Reply:
> simply cast the std::byte* to the underlying type without really needing a column-device-view

Totally makes sense.

After removing dictionaries from the shared memory groupby, it now only handles integers and decimals. A pointer to the underlying type is sufficient for this. Let me unify the shared and global memory aggregators first. I’m really happy it has come down to such a simple solution. Thanks a lot for your input! @davidwendt

Reply:
I've been working on unifying the aggregators locally, and this requires non-trivial effort around performance improvement and API design.

I started with updating null masks to use `bitmask_type*` instead of `bool*`, but I noticed around a 10% slowdown because `bitmask_type` always requires atomic operations to use. After that, there are some differences in behavior between the shared memory aggregator, global memory aggregator, and the existing row aggregator, particularly in how they handle nulls and supported data types. I would suggest that we merge the current PR as is and track the follow-up work through #17032.